worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerQueueConsumerImpl.java

```diff
@@ -34,86 +42,231 @@
+
 /**
  * QueueConsumer implementation for a WorkerQueue. This QueueConsumer hands off messages to worker-core upon delivery assuming the message
  * is not marked 'redelivered'. Redelivered messages are republished to the retry queue with an incremented retry count. Redelivered
  * messages that have exceeded the retry count are republished to the rejected queue.
  */
 public class WorkerQueueConsumerImpl implements QueueConsumer
 {
-    public static final String REJECTED_REASON_TASKMESSAGE = "TASKMESSAGE_INVALID";
+    public static final String REJECTED_REASON_TASKMESSAGE_INVALID = "TASKMESSAGE_INVALID";
+    private static final String REJECTED_REASON_PAYLOAD_OFFLOADING_TASKMESSAGE_DATASTOREE_REFERENCE_NOT_FOUND_ERROR = 
+            "TASKMESSAGE_DATASTORE_REFERENCE_NOT_FOUND_ERROR";
+
     private final TaskCallback callback;
     private final RabbitMetricsReporter metrics;
     private final BlockingQueue<Event<QueueConsumer>> consumerEventQueue;
     private final BlockingQueue<Event<WorkerPublisher>> publisherEventQueue;
     private final Channel channel;
     private final String retryRoutingKey;
     private final int retryLimit;
+    private final String invalidRoutingKey;
+    private final ManagedDataStore dataStore;
+    private final Codec codec;
+    private final Runnable disconnectCallback;
+    private final SortedMap<Long, String> offloadedPayloadsToDelete;
+
     private static final Logger LOG = LoggerFactory.getLogger(WorkerQueueConsumerImpl.class);
 
-    public WorkerQueueConsumerImpl(TaskCallback callback, RabbitMetricsReporter metrics, BlockingQueue<Event<QueueConsumer>> queue, Channel ch,
-                                   BlockingQueue<Event<WorkerPublisher>> pubQueue, String retryKey, int retryLimit)
+    private enum PoisonMessageStatus
     {
+        NOT_POISON,
+        CLASSIC_POISON,
+        POISON
+    }
+
+    public WorkerQueueConsumerImpl(TaskCallback callback, RabbitMetricsReporter metrics, BlockingQueue<Event<QueueConsumer>> queue, Channel ch,
+                                   BlockingQueue<Event<WorkerPublisher>> pubQueue, String retryKey, int retryLimit,
+                                   final String invalidKey,
+                                   final ManagedDataStore dataStore, final Codec codec,
+                                   final Runnable disconnectCallback) {
         this.callback = Objects.requireNonNull(callback);
         this.metrics = Objects.requireNonNull(metrics);
         this.consumerEventQueue = Objects.requireNonNull(queue);
         this.channel = Objects.requireNonNull(ch);
         this.publisherEventQueue = Objects.requireNonNull(pubQueue);
         this.retryRoutingKey = Objects.requireNonNull(retryKey);
         this.retryLimit = retryLimit;
+        this.invalidRoutingKey = invalidKey;
+        this.dataStore = Objects.requireNonNull(dataStore);
+        this.codec = Objects.requireNonNull(codec);
+        this.disconnectCallback = disconnectCallback;
```

Comment:
Using `Objects.requireNonNull()` with `dataStore` and `codec`, but not with `invalidKey` and `disconnectCallback` makes it look like `null` is acceptable for these, which I assume is not the case?